ipa-core/src/sharding.rs

```diff
@@ -167,13 +167,28 @@ pub trait ShardConfiguration {
     }
 }
 
-pub trait ShardBinding: Debug + Send + Sync + Clone + 'static {}
+/// This is a runtime version of `ShardBinding`. It is used by the stream interceptor to
+/// avoid type parameter proliferation. It should not be used by protocols.
+pub type ShardContext = Option<ShardIndex>;
+
+pub trait ShardBinding: Debug + Send + Sync + Clone + 'static {
+    fn context(&self) -> ShardContext;
```

Comment:
Besides the shard index, the shard count is useful for some of the functionality I'm working on.

One option would be for me to expand `ShardContext` to also have the shard count but it will look very similar to `Sharded`. Is there any special additional meaning that `ShardContext` has that `Sharded` doesn't?

If they are interchangeable I could modify the code as follows: 
``` fn context(&self) -> Sharded;``` 
then that makes me wonder why have the context function to begin with?

If we don't need it why not just keep `ShardBinding` as it was, without the `context` function and have the following code?
```impl ShardBinding for Sharded {}```

Reply:
`ShardContext` is `Option<ShardIndex>`, it can represent either `NotSharded` or `Sharded`. `Sharded` is definitely sharded and contains a non-optional shard index (and the shard count).

What are you trying to do with `ShardContext`?

Reply:
Here's what I want to do:

I need shard transports (in particular the in-memory version) to know how many shards are in total. As an example, I use this functionality in the Prepare Query API (query processor) so that the leader shard can send messages to all other shards. Note there's no `Context` at this stage, these are the preliminary steps of the query.

My plan was to modify `TransportConfig`, part of `InMemoryTransport`. This config is now created by `InMemoryMpcNetwork::with_stream_interceptor` using the `ShardContext`.

Reply:
If you haven't already, it might be worth some time thinking about it from the other end -- how does `trait Transport` support determining the shard count? It seems related to `fn identity`, but doesn't directly fit into that. 

I wonder if you can use `HashMap::len` on `InMemoryTransport::connections` to determine the shard count? (Or it might be something more like `peer_count`, since you could call it for an MPC transport as well, although there is little point in doing so, because it's always the same.)

I think it would be reasonable to change `ShardContext` to something like this:
```
enum TransportContext {
    // possibly shard count here, although I'm not sure you need it
    MpcNetwork { shard_index: Option<ShardIndex> }, 

    // possibly helper identity here, by analogy to MpcNetwork
    ShardNetwork { shard_count: ShardIndex },
}
```
I'm not sure how `TransportContext` relates to the existing `identity` and `shard_index` fields. Currently those fields are overloaded in that they have different meanings depending on whether the transport being described is a shard network or an MPC network. Maybe add a `local_identity` field to each of the `TransportContext` variants.

Another possibility here is to add a `ShardBinding`-like type parameter to `InMemoryTransport`. I don't think `Sharded` and `NotSharded` are going to work as-is though.

After this PR, there is a set of three functions on `TransportConfigBuilder`: `with_sharding`, `bind_to_shard`, and `not_sharded`. They should correspond in some manner to the enum variants above -- may make sense to rename these functions as well.